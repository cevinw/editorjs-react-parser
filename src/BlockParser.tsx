import React from "react";
import AlertBlock, { AlertConfig } from "./BlockParsers/alert";
import ParagraphBlock, { ParagraphConfig } from "./BlockParsers/paragraph";
import CodeBlock, { CodeConfig } from "./BlockParsers/code";
import DelimiterBlock, { DelimiterConfig } from "./BlockParsers/delimiter";
import HeaderBlock from "./BlockParsers/header";
import ImageBlock, { ImageConfig } from "./BlockParsers/image";
import ColumnsBlock, { ColumnsConfig } from "./BlockParsers/columns";
import QuoteBlock, { QuoteConfig } from "./BlockParsers/quote";
import TableBlock, { TableConfig } from "./BlockParsers/table";
import ListBlock, { ListConfig } from "./BlockParsers/list";
import ErrorBlock, { ErrorConfig } from "./BlockParsers/error";
import Raw from "./BlockParsers/raw";

export interface BlockParserProps {
  data: OutputData;
  config?: BlockParserConfig;
}

export interface BlockParserConfig {
  alert?: AlertConfig;
  code?: CodeConfig;
  columns?: ColumnsConfig;
  table?: TableConfig;
  image?: ImageConfig;
  paragraph?: ParagraphConfig;
  list?: ListConfig;
  quote?: QuoteConfig;
  delimiter?: DelimiterConfig;
  error?: ErrorConfig;
}

export interface OutputBlockData<Data extends object = any> {
  id?: BlockId;
  type: string;
  data: BlockToolData<Data>;
  tunes?: { [name: string]: BlockTuneData };
}

export interface OutputData {
  version?: string;
  time?: number;
  blocks: OutputBlockData[];
}

export type BlockId = string;
export type BlockToolData<T extends object = any> = T;
export type BlockTuneData = any;

export const BlockParser = ({ data, config }: BlockParserProps) =>
  data.blocks.map((blockData: OutputBlockData) => parseData(blockData, config));

const parseData = (
  item: OutputBlockData,
  config?: BlockParserConfig,
): React.JSX.Element => {
  return (
    {
      raw: <Raw key={item.id} item={item} />,
      header: <HeaderBlock key={item.id} item={item} />,
      code: <CodeBlock key={item.id} item={item} config={config?.code} />,
      list: <ListBlock key={item.id} item={item} config={config?.list} />,
      alert: <AlertBlock key={item.id} item={item} config={config?.alert} />,
      quote: <QuoteBlock key={item.id} item={item} config={config?.quote} />,
      table: <TableBlock key={item.id} item={item} config={config?.table} />,
      image: <ImageBlock key={item.id} item={item} config={config?.image} />,
      delimiter: (
        <DelimiterBlock key={item.id} item={item} config={config?.delimiter} />
      ),
      paragraph: (
        <ParagraphBlock key={item.id} item={item} config={config?.paragraph} />
      ),
      columns: (
        <ColumnsBlock
          key={item.id}
          item={item}
          config={config?.columns}
          blockRendererConfig={config}
        />
      ),
    }[item.type] || (
      <ErrorBlock key={item.id} item={item} config={config?.error} />
    )
  );
};

export default BlockParser;
