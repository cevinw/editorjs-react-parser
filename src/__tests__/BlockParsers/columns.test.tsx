import { describe, expect } from "vitest";
import { render } from "@testing-library/react";
import React from "react";
import ColumnsBlock, { EditorJsColumns } from "../../BlockParsers/columns";
import { OutputBlockData } from "../../BlockParser";
import test from "../conftest";

describe("Column block tests", () => {
  test("Should parse column block", async ({ data }) => {
    const outputData = await data;
    const inputItem: OutputBlockData<EditorJsColumns> = {
      id: "testid",
      type: "columns",
      data: {
        cols: [outputData.data, outputData.data],
      },
    };
    const { container } = render(<ColumnsBlock item={inputItem} />);
    expect(container).toMatchSnapshot();
  });
});
