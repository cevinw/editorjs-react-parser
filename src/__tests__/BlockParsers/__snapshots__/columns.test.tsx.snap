// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Column block tests > Should parse column block 1`] = `
<div>
  <section
    class="my-2 md:grid md:gap-4 md:grid-cols-2"
  >
    <div
      class="self-center"
    >
      <section
        class="grid mt-2"
      >
        <img
          alt=""
          class="rounded justify-self-center dark:dark:opacity-75"
          src="http://test.test.png"
          style="width: 500px;"
        />
      </section>
      <p
        class="mt-1"
      >
        test
      </p>
      <figure
        class="text-sm rounded-md overflow-hidden shadow-sm mt-2"
      >
        <pre
          style="display: block; overflow-x: auto; padding: 0.5em; background: rgb(35, 35, 35); color: rgb(230, 225, 220);"
        >
          <code
            style="white-space: pre;"
          >
            <span
              style="color: rgb(109, 156, 190);"
            >
              test
            </span>
          </code>
        </pre>
      </figure>
      <ol>
        <li
          class="list-decimal ml-7"
        >
          test
        </li>
        <li
          class="list-decimal ml-7"
        >
          test2
        </li>
      </ol>
      <p
        class="text-left"
      >
        <span
          class="italic text-lg"
        >
          "
          test
          "
        </span>
        <br />
        <span
          class="opacity-85 text-sm"
        >
          - 
          test
        </span>
      </p>
      <p
        class="mt-1"
      >
        test
      </p>
      <section
        class="my-2 md:grid md:gap-4 md:grid-cols-2"
      >
        <div
          class="self-center"
        >
          <p
            class="mt-1"
          >
            test
          </p>
        </div>
        <div
          class="self-center"
        >
          <p
            class="mt-1"
          >
            test
          </p>
        </div>
      </section>
      <table
        class="w-full"
      >
        <tbody>
          <tr>
            <th
              class="first:border-l-0 text-left pl-2 last:border-r-0 border-b-2 border"
            >
              test
            </th>
            <th
              class="first:border-l-0 text-left pl-2 last:border-r-0 border-b-2 border"
            >
              test
            </th>
          </tr>
          <tr>
            <td
              class="first:border-l-0 last:border-r-0 pl-2 border-b border"
            >
              test
            </td>
            <td
              class="first:border-l-0 last:border-r-0 pl-2 border-b border"
            >
              test
            </td>
          </tr>
        </tbody>
      </table>
      <figure
        class="mt-2 p-2 mt-2 px-4 flex bg-opacity-50 shadow-sm rounded-lg mt-2 bg-gray-300 text-gray-600"
      >
        <figcaption
          class="text-right"
        >
          test
        </figcaption>
      </figure>
      <h2>
        test
      </h2>
    </div>
    <div
      class="self-center"
    >
      <section
        class="grid mt-2"
      >
        <img
          alt=""
          class="rounded justify-self-center dark:dark:opacity-75"
          src="http://test.test.png"
          style="width: 500px;"
        />
      </section>
      <p
        class="mt-1"
      >
        test
      </p>
      <figure
        class="text-sm rounded-md overflow-hidden shadow-sm mt-2"
      >
        <pre
          style="display: block; overflow-x: auto; padding: 0.5em; background: rgb(35, 35, 35); color: rgb(230, 225, 220);"
        >
          <code
            style="white-space: pre;"
          >
            <span
              style="color: rgb(109, 156, 190);"
            >
              test
            </span>
          </code>
        </pre>
      </figure>
      <ol>
        <li
          class="list-decimal ml-7"
        >
          test
        </li>
        <li
          class="list-decimal ml-7"
        >
          test2
        </li>
      </ol>
      <p
        class="text-left"
      >
        <span
          class="italic text-lg"
        >
          "
          test
          "
        </span>
        <br />
        <span
          class="opacity-85 text-sm"
        >
          - 
          test
        </span>
      </p>
      <p
        class="mt-1"
      >
        test
      </p>
      <section
        class="my-2 md:grid md:gap-4 md:grid-cols-2"
      >
        <div
          class="self-center"
        >
          <p
            class="mt-1"
          >
            test
          </p>
        </div>
        <div
          class="self-center"
        >
          <p
            class="mt-1"
          >
            test
          </p>
        </div>
      </section>
      <table
        class="w-full"
      >
        <tbody>
          <tr>
            <th
              class="first:border-l-0 text-left pl-2 last:border-r-0 border-b-2 border"
            >
              test
            </th>
            <th
              class="first:border-l-0 text-left pl-2 last:border-r-0 border-b-2 border"
            >
              test
            </th>
          </tr>
          <tr>
            <td
              class="first:border-l-0 last:border-r-0 pl-2 border-b border"
            >
              test
            </td>
            <td
              class="first:border-l-0 last:border-r-0 pl-2 border-b border"
            >
              test
            </td>
          </tr>
        </tbody>
      </table>
      <figure
        class="mt-2 p-2 mt-2 px-4 flex bg-opacity-50 shadow-sm rounded-lg mt-2 bg-gray-300 text-gray-600"
      >
        <figcaption
          class="text-right"
        >
          test
        </figcaption>
      </figure>
      <h2>
        test
      </h2>
    </div>
  </section>
</div>
`;
